[xcpmd] Updates for acpid2, multi-battery support and general code cleanup

jctemkin opened the pull request:
Pursuant to [OXT-342](https://openxt.atlassian.net/browse/OXT-342), here are some changes and improvements for xcpmd, both related to the adoption of acpid2 and otherwise. 
- Netlink-related code pruned, since acpid2 forwards netlink messages
- Ability to listen for lid, AC adapter, power/suspend/sleep/brightness buttons, and tablet mode events via acpid
- Battery events now event-driven instead of polling
- A generalized battery interface that supports >2 batteries
- Reorganization of xenstore nodes--battery nodes are now nested under /pm/batX and /pm/events/batX, though old structures are maintained for compatibility
- Safer ACPI event string parsing
- Cleanup of leftover thermal code
- General cleanup of battery code
- General cleanup of prototypes.h

Commit:
Updates for acpid2, better multi-battery support and general code
cleanup.

Commit:
Add back support for current_battery_level.

Commit:
Remove some leftover debug code.

jean-edouard commented:
I didn't read the code yet, but I have a concern about battery updates coming from acpid (I actually have an outstanding pull request about this, that's probably conflicting with this one, https://github.com/OpenXT/xctools/pull/12).

My original concern is that at the point when xcpmd gets a "battery status changed signal" from acpid, /sys hasn't been updated yet, and xcpmd (and programs listening to it) will go read that stale information from /sys.
Your pull request may already address that somehow, in which case I'd like to know more!

Thanks

jctemkin commented:
Unless there's some multi-threaded magic going on that I'm missing, the [driver code](https://github.com/torvalds/linux/blob/v3.18-rc7/drivers/acpi/battery.c#L1081) updates the battery status in /sys before issuing the netlink signal. My understanding is that the reason the stale status was a problem was because xcpmd was only updating the status at a 4-second interval, but was forwarding the battery status notifications as they occurred. 

My pull request addresses this by using the ACPI events themselves to trigger an update of its own information, and only after that does it send a notification on dbus, which I believe is just an event-driven version of #12 anyway.

jean-edouard commented:
Oh that makes sense, I didn't realize that, thanks for clarifying!
I'll close #12.

rossphilipson commented:
Because all the battery code as moved in the same commit where it was changed, it is very difficult to see what was changed. Would it be hideously difficult to rework it so one commit has the changes and the final commit has the big move?

Commit:
Whoops forgot to prune the netlink files.

jctemkin commented:
I made a branch with one commit for [code changes](https://github.com/jctemkin/xctools/commit/947dd7b77d9550c49912a7e2e384f86d01399ae9), and another for [code rearrangement](https://github.com/jctemkin/xctools/commit/b042eafcb50a92a944cfc2444429d79dc0410579)--hopefully that makes things easier to read.

rossphilipson commented:
So it looks like a lot of it is just reformatting all the code?

rossphilipson commented:
Sorry I accidentally hit the button to close it.

jctemkin commented:
There's a good amount of reformatting and rearrangement, but there's also fairly significant changes in the way ACPI messages are processed and in how battery info/status is updated.

rossphilipson commented:
Sorry I did not mean to imply it was all reformatting; I know you did a bunch of good work in there in the ACPI event processing etc. I guess I was just surprised by the amount of reformatting.

jean-edouard commented:
Thanks for your readability commits, though next time please modify the pull request itself, so that even after the merge people can still go back and clearly see which aspects of the logic got modified.

Reviewing the code using those new commits is definitely easier, even if there's still a fair amount of refactoring in the "code changes" commit. Most of it makes sense, like removing tabs. One thing that bothers me a little bit though, is the fact that you didn't follow the comments format "/\* */" used all over the code, but introduced the "//" notation.
I guess it's acceptable for the comments you added, but you actually changed some comments from their original notation to the one you introduced, without changing their content. In git, you wrongfully end up becoming the author of said comments, which is misleading and not cool for the actual author...

No big deal for this PR, next time please just try to avoid unnecessary edits...

Apart from that, this builds fine, @rossphilipson and/or I will try to go through more of the code and do some basic testing.

jctemkin commented:
Duly noted; I'll avoid modifying comments in the future. Thanks for looking this over.

jean-edouard commented:
I did a quick test to validate the timing thing we discussed previously.
On an install of a build that includes your changes, I opened 2 terminals.
In the first terminal I run `dbus-monitor --system | grep battery_status_changed`.
In the second terminal, I type `xec -s com.citrix.xenclient.xcpmd battery-state 0`.
Then I plug/unplug the power, and as soon as the first terminal outputs a line, I run the command in the second terminal a bunch of times.

I got varying results... Most of the time, after the first or second dbus signal, the battery state is correctly updated. However, sometimes, when going from unplugged to plugged, only one signal happens and the battery status never goes from unplugged to plugged.
That is until you unplug the AC again. At that point, the battery state goes briefly from unplugged to plugged, and then to unplugged. It's as if the last event got stuck in the queue until another one came in.

The previous paragraph being quite unreadable, here's a "diagram" of the battery state over time in the failing case:
[AC plugged] 4 4 4 4 [AC unplugged] 4 2 2 2 [AC plugged] 2 2 2 2 [AC unplugged] 4 2 2 2
- The "4" after the first unplug is expected, an event comes just after to fix it.
- After the AC got plugged, "2" is expected once or twice, but actually stays forever.
- When re-unplugging, the previously expected "4" finally arrives, just before the expected "2".

Sorry for the lengthy comment, this is pretty difficult to explain...

rossphilipson commented:
I think it would be better to put defect findings like this in the ticket and reference the PR. I could (effectively) disappear if say the PR is closed out.

jean-edouard commented:
Good idea, done.

Commit:
Change battery code back to polling because of unreliable hardware.

jean-edouard commented:
Tested. I keep getting the signal even when nothing happens.
After a quick look at the code, I think you used !memcmp instead of just memcmp...

Commit:
Actually check for changes

jctemkin commented:
Right you are. Sigh... hopefully fixed now.

jean-edouard commented:
One more question, how do you handle the case where there's 3 batteries slots, with batteries in slots 0 and 2, and no battery in slot 1?

jctemkin commented:
When update_batteries() is called, it checks the sysfs for every battery up to MAX_BATTERIES_SUPPORTED, and tracks their information individually. There shouldn't be a reason that the absence of a slot should be handled any differently whether it's between two present slots or not, but I admit I don't have a three-battery platform to test that supposition.

jean-edouard commented:
It's a situation that we've seen before, and that is handled by the current code.
We want to maximize the amount of battery information we have, and not shutdown if battery 0 is empty when there's a full battery 2.
Losing that feature would be a regression...

Ultimately we'd want to generate guest ACPI code on the fly depending on the hardware, and support an infinite number of batteries, but right now the idea is to get the best out of our 2 static "emulated" batteries.

jctemkin commented:
Ahh, I see what you're getting at. The behavior of the bst1/bif1 xenstore nodes is preserved for that case.

jean-edouard commented:
Do you mean the xenstore node will handle those two batteries but not the dbus interface?
We definitely want consistency between the UI (dbus) and the guests (xenstore).

Commit:
Add batteries_present() DBus method for UIVM use.

Commit:
Handle battery slot removal gracefully.

Commit:
Support an arbitrary number of batteries and add DBus bits for UIVM.

Commit:
Fix bug where directories were opened but not closed.

Commit:
Fix case where no "num_batteries_changed" signal is emitted when one battery is removed and a different one inserted in the same 4-second interval.

jctemkin commented:
I've modified the battery model to support an arbitrary number of batteries, and added a DBus method to communicate what batteries are present and a signal when the number of batteries present changes. I believe that should be all the information the UIVM needs for its battery applet.

jean-edouard commented:
I gave this a quick test on a laptop with 3 battery slots (BAT1 not being wired to anything), and got the correct bif/bst nodes. Also noticed bif1 and bst1 disappeared when I unplugged BAT2, which is what's expected.
More to come next week.

jean-edouard commented on xcpmd/src/acpi-events.c:
```diff
@@ -30,133 +30,406 @@
+    write_battery_info_to_xenstore(battery_index);
+
+    //sprintf(path, "%s%i/%s", XS_BATTERY_PATH, battery_index, XS_BATTERY_PRESENT_LEAF);
+    //xenstore_write((last_info[battery_index].present == YES) ? "1" : "0", path);
```
Why is this commented out? Can you maybe explain in a comment or remove those 2 lines?

jctemkin replied:
Those two moved into write_battery_status_to_xenstore(); I missed them during pruning. These can be removed.

jean-edouard commented on xcpmd/src/acpi-events.c:
```diff
@@ -30,133 +30,406 @@
+
+    xenstore_write(lid_status == LID_OPEN ? "open" : "closed", XS_LID_STATE_PATH);
+    xenstore_write("1", XS_LID_EVENT_PATH);
+    //notify_com_citrix_xenclient_xcpmd_lid_changed(xcdbus_conn, XCPMD_SERVICE, XCPMD_PATH);
```
Same here.

jctemkin replied:
This call doesn't yet exist in the IDL. I can go ahead and add it to the [existing pull request](https://github.com/OpenXT/idl/pull/7), if you think we stand to benefit from it.

jean-edouard commented on xcpmd/src/acpi-events.c:
```diff
@@ -30,133 +30,406 @@
+
+    xcpmd_log(LOG_INFO, "Suspend button pressed event\n");
+    xenstore_write("1", XS_SUSPBTN_EVENT_PATH);
+    //notify_com_citrix_xenclient_xcpmd_suspend_button_pressed(xcdbus_conn, XCPMD_SERVICE, XCPMD_PATH);
```
Same here.

jctemkin replied:
Another case of the call not yet existing in the IDL.

jean-edouard commented on xcpmd/src/acpi-events.c:
```diff
@@ -30,133 +30,406 @@
+    }
+    else if (cmd == BCL_CYCLE) {
+        //xenstore_write("3", XS_BCL_CMD);
+        //xenstore_write("1", XS_BCL_EVENT_PATH);
```
And here.

jctemkin replied:
The new version of acpid has the ability to listen for a BCL_CYCLE event, but I wasn't sure exactly who would be consuming the information and if they would support the new event. This was left as a note that the capability is there if we want to go ahead and implement support for this on the other end.

jean-edouard replied:
Adding a short comment there would make it more explicit.

jean-edouard commented on xcpmd/src/acpi-events.c:
```diff
@@ -30,133 +30,406 @@
+        }
     }
+    //else
+    //    xcpmd_log(LOG_DEBUG, "Unhandled ACPI class: %s\n", class);
```
And here.

jctemkin replied:
This was clogging up /var/log/messages something fierce, and not providing any useful debug information at this time. I can go ahead and remove this one.

jean-edouard commented on xcpmd/src/acpi-events.c:
```diff
@@ -166,64 +439,56 @@ void acpi_events_read(void)
 #ifdef XCPMD_DEBUG
         acpi_buffer[len] = '\0';
-        xcpmd_log(LOG_DEBUG, "~ACPI-event: %s\n", acpi_buffer);
+        //xcpmd_log(LOG_DEBUG, "~ACPI-event: %s\n", acpi_buffer);
```
If we're in debug mode, don't we want that debug message?

jctemkin replied:
The answer for my particular use-case was "rarely," but I can go ahead and uncomment it.

jean-edouard commented on xcpmd/src/acpi-events.h:
```diff
@@ -24,14 +24,111 @@
+#define ACPI_PROCESSOR_NOTIFY_POWER         0x81
+#define ACPI_PROCESSOR_NOTIFY_THROTTLING    0x82
+#define ACPI_PROCESSOR_FILE_PERFORMANCE     "performance"
+#define ACPI_PROCESSOR_NOTIFY_PERFORMANCE   0x80
```
Isn't this already defined a few lines above?

jctemkin replied:
Right you are. Away it goes.

jean-edouard commented on xcpmd/src/acpi-events.h:
```diff
@@ -24,14 +24,111 @@
+#define ACPI_SBS_BLOCK_MAX              32
+
+// From drivers/acpi/scan.c
+#define ACPI_BUS_CLASS                   "system_bus"
```
The indent is slightly off.

jean-edouard commented on xcpmd/src/battery.c:
```diff
@@ -0,0 +1,803 @@
+        //Battery directory does not exist--this normally occurs when a battery slot is removed
+        if (errno == ENOENT) {
+            status.present = NO;
+            //memcpy(&last_status[battery_index], &status, sizeof(struct battery_status));
```
?

jctemkin replied:
...that line is definitely not supposed to be commented out. Fixing.

jean-edouard commented on xcpmd/src/battery.c:
```diff
@@ -0,0 +1,803 @@
+
+            fclose(file);
+
+            //Trim off leading spaces.
```
Have you seen values with leading spaces or is this just prevention?

jctemkin replied:
I haven't, but the previous behavior was to trim leading spaces.

jean-edouard replied:
Oh, that must come from the time we were using /proc!
I guess it doesn't hurt to keep doing it.

jean-edouard commented on xcpmd/src/battery.c:
```diff
@@ -0,0 +1,803 @@
+
+    closedir(battery_dir);
+    memcpy(&last_status[battery_index], &status, sizeof(struct battery_status));
+    //print_battery_status();
```
Should this be un-commented for debugging? Why not using `#ifdef XCPMD_DEBUG`?

jctemkin replied:
I had it commented out because the frequent updates were clogging the system log, but I can uncomment it for others' use.

jean-edouard commented on xcpmd/src/battery.c:
```diff
@@ -0,0 +1,803 @@
+
+
+//Gets a battery's info from the sysfs and stores it in last_info.
+int update_battery_info(unsigned int battery_index) {
```
So we read all the /sys nodes twice? Should this function be merged with the previous one in the future?

jctemkin replied:
This is a leftover from the event-driven code. With the current architecture, it would definitely be more efficient to merge this function with the previous one.

jean-edouard commented on xcpmd/src/battery.c:
```diff
@@ -0,0 +1,803 @@
+    }
+
+#ifdef XCPMD_DEBUG
+    //xcpmd_log(LOG_DEBUG, "~Updated battery information in xenstore\n");
```
?

jctemkin replied:
Another bit I commented out during my own debugging... I'll uncomment it for everyone else.

jean-edouard commented on xcpmd/src/xcpmd.h:
```diff
@@ -147,14 +167,29 @@ struct battery_status {
-#define MAX_BATTERY_SUPPORTED               0x2
-#define MAX_BATTERY_SCANNED                 0x5
+//#define MAX_BATTERY_SUPPORTED               0x2
+//#define MAX_BATTERY_SCANNED                 0x5
```
That can probably just go...

jctemkin replied:
Right you are.

jean-edouard commented:
I tested some laptops with interesting battery configurations, and battery-related things worked as expected.

I also did another quick overview of the code, and didn't find anything critical, mainly just wondering about the commented out code.

Commit:
Fix bug on battery removal, and clean up some comment/formatting.

jctemkin commented:
Great to hear, and thank you for looking this over. I added a commit to address the issues you brought up.

jean-edouard commented:
+1

jean-edouard commented:
Did you create a ticket for switching qemu to the new xenstore nodes?
What about an "improvement" ticket for merging the info and status functions?
Thanks

Commit:
Add comment explaining BCL_CYCLE section.

jctemkin commented:
[Done](https://openxt.atlassian.net/browse/OXT-366) and [done](https://openxt.atlassian.net/browse/OXT-367).

jean-edouard commented:
Thank you!

cjp256 commented on xcpmd/src/acpi-events.c:
```diff
@@ -30,133 +30,403 @@
+    update_battery_info(battery_index);
+    write_battery_info_to_xenstore(battery_index);
+
+    sprintf(path, "%s%i/%s", XS_BATTERY_EVENT_PATH, battery_index, XS_BATTERY_INFO_EVENT_LEAF);
```
Better to use snprintf over sprintf (and throughout)

cjp256 commented on xcpmd/src/acpi-events.c:
```diff
@@ -30,133 +30,403 @@
+    else if (file == NULL)
+        return AC_UNKNOWN;
+
+    fgets(data, sizeof(data), file);
```
are you guaranteed to get all of the data desired on fgets? and be null terminated?

jctemkin replied:
Assuming the battery driver doesn't change radically, this fgets will retrieve all the data desired for the files we're interested in. It does automatically null-terminate.

cjp256 commented on xcpmd/src/battery.c:
```diff
@@ -0,0 +1,805 @@
+        info->design_voltage = strtoull(attrib_value, NULL, 10) / 1000;
+    }
+    else if (!strcmp(attrib_name, "model_name")) {
+        strncpy(info->model_number, attrib_value, 32);
```
where does 32 come from? can that be a sizeof()? (and throughout)

cjp256 commented on xcpmd/src/battery.c:
```diff
@@ -0,0 +1,805 @@
+    memset(bif, 0, 1024);
+    memset(string_info, 0, 256);
+    // write 9 dwords (so 9*4) + length of 4 strings + 4 null terminators
+    snprintf(bif, 3, "%02x",
```
possible to overflow into 3 digits and then lose a digit?

cjp256 commented:
looks pretty good to me, nice job! i realize some of my comments are on original sources that showed up in the diff due to refactoring, so not necessarily new specific to this PR.

Commit:
Change sprintfs to snprintfs.

jctemkin commented:
All sprintfs are now snprintfs. The other comments regarded original sources, so I left those alone--I can go ahead and stick some sizeof()s in there if you think it's warranted, though.

stacktrust commented:
If you have the time to bring the original sources up to the same security standard as this PR, including sections which don't appear in these diffs, could that be a new PR?

jctemkin commented:
I don't have the cycles to take care of that myself at the moment, but you're welcome to create a new Jira ticket for an improvement.

jean-edouard commented:
I think this is ready to be merged. I will do so later today.
